Source/SwiftLintCore/Models/Baseline.swift

```diff
@@ -0,0 +1,169 @@
+import Foundation
+
+private typealias GroupedViolations = [String: [BaselineViolation]]
+
+private struct BaselineViolation: Equatable, Codable, Hashable {
+    let violation: StyleViolation
+    let text: String
+    var key: String { text + violation.reason + violation.severity.rawValue }
+
+    init(violation: StyleViolation, text: String) {
+        let location = violation.location
+        self.violation = violation.with(location: Location(
+            // Within the baseline, we use relative paths, so that
+            // comparisons are independent of the absolute path
+            file: location.relativeFile,
+            line: location.line,
+            character: location.character)
+        )
+        self.text = text
+    }
+}
+
+/// A set of violations that can be used to filter newly detected violations.
+public struct Baseline: Equatable {
+    private let baselineViolations: GroupedViolations
+
+    /// The stored violations.
+    public var violations: [StyleViolation] {
+        baselineViolations.keys.sorted().flatMap({ baselineViolations[$0]! }).violationsWithAbsolutePaths
+    }
+
+    /// Creates a `Baseline` from a saved file.
+    ///
+    /// - parameter fromPath: The path to read from.
+    public init(fromPath path: String) throws {
+        let data = try Data(contentsOf: URL(fileURLWithPath: path))
+        baselineViolations = try JSONDecoder().decode(GroupedViolations.self, from: data)
+    }
+
+    init(violations: [StyleViolation]) {
+        baselineViolations = violations.baselineViolations.groupedByFile()
+    }
+
+    /// Writes a `Baseline` to disk in JSON format.
+    ///
+    /// - parameter violations: The violations to save.
+    /// - parameter toPath: The path to write to.
+    public static func write(_ violations: [StyleViolation], toPath path: String) throws {
+        try write(violations.baselineViolations.groupedByFile(), toPath: path)
+    }
+
+    private static func write(_ violations: GroupedViolations, toPath path: String) throws {
+        let data = try JSONEncoder().encode(violations)
+        try data.write(to: URL(fileURLWithPath: path))
+    }
+
+    /// Filters out violations that are present in the `Baseline`.
+    ///
+    /// Assumes that all violations are from the same file.
+    ///
+    /// - parameter violations: The violations to filter.
+    /// - Returns: The new violations.
+    public func filter(_ violations: [StyleViolation]) -> [StyleViolation] {
+        guard let firstViolation = violations.first,
+              let baselineViolations = baselineViolations[firstViolation.location.relativeFile ?? ""],
+              baselineViolations.isNotEmpty else {
```

Comment:
Please help me to understand this. Why is only the first violation checked for existence in the baseline?

Reply:
So we want to check whether there are any incoming violations, and whether we have any violations in the baseline for the file. When called by the LintOrAnalyzeCommand, the method will be passed the list of violations for a specific file ("Assumes that all violations are from the same file."), so we only need to look at the file for the first violation to look up the baseline violations for the file.